File: java_target/src/FoxySheep.java
//import java.io.FileInputStream;
//import java.io.InputStream;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

//import org.antlr.v4.runtime.ANTLRInputStream;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.ParseTreeWalker;


public class FoxySheep {

    // Cache the utility objects.
	private static FoxySheepParser parser;
	private static FoxySheepLexer lexer;
	private static FullFormParser ff_parser;
	private static FullFormLexer ff_lexer;
	private static FullFormEmitter emitter;
	private static ParseTreeWalker walker;
	private static PostParser postParser;

	public static ParseTree postParse(ParseTree tree){
        // Reuse existing objects.
	    if(walker == null) walker = new ParseTreeWalker();
        if(postParser == null) postParser = new PostParser();

		walker.walk(postParser, tree);
		if(tree.getParent() != null){
			tree = tree.getParent();
		}
		return tree;
	}

	public static ParseTree parseTreeFromString(String input){
	    //Reuse any existing parser or lexer.
	    if(lexer == null) {
	        lexer = new FoxySheepLexer(CharStreams.fromString((input)));
        } else{
	        lexer.setInputStream(CharStreams.fromString((input)));
        }
        if(parser == null) {
	        parser = new FoxySheepParser(new CommonTokenStream(lexer));
        } else{
	        parser.setTokenStream(new CommonTokenStream(lexer));
        }

        ParseTree tree = parser.prog();
        tree = postParse(tree);
		return tree;
	}

	public static ParseTree ffParseTreeFromString(String input){
        //Reuse any existing parser or lexer.
        if(ff_lexer == null) {
            ff_lexer = new FullFormLexer(CharStreams.fromString((input)));
        } else{
            ff_lexer.setInputStream(CharStreams.fromString((input)));
        }
        if(ff_parser == null) {
            ff_parser = new FullFormParser(new CommonTokenStream(ff_lexer));
        } else{
            ff_parser.setTokenStream(new CommonTokenStream(ff_lexer));
        }

        //Parse!
        return ff_parser.prog();
	}

	public static String fullFormFromString(String input){
	    // Reuse existing emitter.
        if(emitter == null){
            emitter = new FullFormEmitter();
        }

        // Parse the input.
        ParseTree tree = parseTreeFromString(input);

        // Emit FullForm
		return emitter.visit(tree);
	}

	public static void main(String[] args) throws Exception {
        BufferedReader cin = new BufferedReader(new InputStreamReader(System.in));
        String user_in;

        // Simple REPL
        while(true){
            System.out.print("in:= ");
            user_in = cin.readLine();
            if(user_in.isEmpty()){
                break;
            }
            System.out.println(fullFormFromString(user_in));
        }
	}

}

File: python_target/FoxySheep.py
from antlr4 import *

from FoxySheep import FoxySheepLexer
from FoxySheep import FoxySheepParser
from FoxySheep import FullFormEmitter
from FoxySheep import PostParser
from FoxySheep import FullFormLexer
from FoxySheep import FullFormParser

# Cache the utility objects.
parser = None
lexer = None
ff_parser = None
ff_lexer = None
emitter = None


def postParse(tree):
    """Post process the parse tree. In particular, flatten some flat
    operators. Some operators appear in the source text without any
    explicit associativity, such as `Plus`, and are parsed into arbitrary
    tree structures.
    """

    walker = ParseTreeWalker()
    post_parser = PostParser()
    walker.walk(post_parser, tree)

    # The PostParser can restructure the tree in a way that changes the root.
    if tree.parentCtx is not None:
        tree = tree.parentCtx
    return tree


def parse_tree_from_string(input:str, post_process=True):

    # Boilerplate
    # lexer = FoxySheepLexer(InputStream(input))
    # stream = CommonTokenStream(lexer)
    # parser = FoxySheepParser(stream)

    global parser, lexer

    # Reuse any existing parser or lexer.
    if not lexer:
        lexer = FoxySheepLexer(InputStream(input))
    else:
        lexer.inputStream = InputStream(input)
    if not parser:
        parser = FoxySheepParser(CommonTokenStream(lexer))
    else:
        parser.setTokenStream(CommonTokenStream(lexer))

    # Parse!
    tree = parser.prog()
    if post_process:
        tree = postParse(tree)
    return tree


def ff_parse_tree_from_string(input:str):
    global ff_parser, ff_lexer

    # Reuse any existing parser or lexer.
    if not ff_lexer:
        ff_lexer = FullFormLexer(InputStream(input))
    else:
        ff_lexer.InputStream = InputStream(input)
        ff_lexer.reset()
    if not ff_parser:
        stream = CommonTokenStream(ff_lexer)
        ff_parser = FullFormParser(stream)
    else:
        # Don't create a new parser. Just reset the input stream.
        ff_parser.setTokenStream(CommonTokenStream(ff_lexer))
        ff_parser.reset()

    # Parse!
    return ff_parser.prog()


def FullForm_from_string(input:str):
    global emitter

    # Reuse existing emitter.
    if not emitter:
        emitter = FullFormEmitter()

    # Parse the input.
    tree = parse_tree_from_string(input)

    # Emit FullForm.
    return emitter.visit(tree)


def main():
    #Simple REPL
    while True:
        try:
            user_in = input('in:= ')
        except KeyboardInterrupt:
            break
        if user_in == '':
            break

        print(FullForm_from_string(user_in))


if __name__ == '__main__':
    main()
